### ISSUE TITLE
No way to use custom options in Ruby

### ISSUE THREAD

**Issue by kailan** [protocolbuffers/protobuf#1198]
I created a custom option extension that I use in my proto files, like this:

``` proto
service Fishtank {
  rpc GetFish(FishQuery) returns (Fish) {
    option (protapi.http).get = "/fish/{id}";
  }
}
```

It doesn't seem that once it is compiled into Ruby code, I can access the value of the `protapi.http` option. I believe this can be done in other languages by accessing the descriptor.

Am I missing something, or is this not possible right now?

**Comment by haberman** [comment]
You are correct that Ruby does not yet support custom options. Hopefully we can add this before too long.

**Comment by kblanks** [comment]
@haberman Awesome! It'll really help out a project of mine.

Any way I can help out with implementing this?

**Comment by skipperguy12** [comment]
This would be helpful for me as well

**Comment by haberman** [comment]
This is becoming a more and more common request. I have been thinking through how the implementation ought to work on this. It's slightly tricky to modify the DSL for this, because custom options let you declare the option and use it both in the same file. This is a challenge to accommodate in a clean way.

**Comment by kblanks** [comment]
Any progress on this? @haberman

**Comment by haberman** [comment]
Ok there are a couple parts to making this work.

First of all we need to support proto2 field presence. Right now the code generator fails immediately if it sees a file with `syntax="proto2";`. Step 1 would be a PR that supports proto2 files that don't contain any extensions or custom options. Proto2 "optional" fields should be supported by adding hasbits to the in-memory representation to indicate whether a field is present or not. I'm happy to explain this in more detail for anyone who wants to implement this.

Secondly we need to support proto2 extensions. This means extending the DSL with something like `add_extension` and making it so you can read and write extension fields of extended messages with something like `msg[my_extension] = 5`.

Finally we need a way of extending the DSL to support custom options. We can talk more about this part once #1 and #2 are done.

I'd be happy to see pull requests for any/all of these, and I'd be happy to help guide people who want to implement this to the right places in the code.

**Comment by kblanks** [comment]
I'd love to work on this but I have one question: Why would proto2 need to be supported? Surely implementing this for proto3 would be enough.

**Comment by haberman** [comment]
Custom options are defined as extensions of messages `descriptor.proto`, which is a `syntax="proto2";` file. Extensions don't exist at all in proto3.

So even if your messages are only proto3, we need proto2 support to be able to represent the options themselves.

**Comment by kblanks** [comment]
Ah, makes sense. I'll give it a shot later on.

**Comment by nerdrew** [comment]
Any status updates on proto2 support?

**Comment by djudd-stripe** [comment]
Any progress since last year? I might have a few days to push this forward.

**Comment by nerdrew** [comment]
Funny. We were thinking of looking at this soon too. cc @zanker

**Comment by zanker** [comment]
Hah, @djudd-stripe I was actually talking to @nerdrew about this last week and we assumed that Stripe didn't use protobufs. Are you using the "protobuf" gem currently?

I was able to hack something to support defaults in a few days, so I don't think it's too hard to at least get basic protobuf2 support at least. It looks like most of the work is building out the Ruby <-> C bridges, as UPB handles everything already.

I'm still working on proving this out internally to make sure we can feasibly swap to this, but right now we're interested in adding at least basic synta2 support.

@haberman it sounds like on your end, you would be interested in merging syntax2 support to the Ruby gem, as long as it meets code standards and is done properly?

**Comment by djudd-stripe** [comment]
Not a lot of Protobuf at Stripe right now but we are exploring adding more. We're using this code so far but are interested in custom annotations.

**Comment by djudd-stripe** [comment]
@zanker it sounds like you have Proto2 support work in-progress? Anything more useful I can do than wait for you to get it merged, then maybe add custom option support on top (unless you've included that already)?

**Comment by zanker** [comment]
@djudd-stripe This is still very much WIP (hence the random debug statements), but for reference: https://github.com/google/protobuf/compare/master...zanker:zanker/syntax2 is what things are looking like. I don't think wiring in extensions would be _too_ hard overall.

**Comment by pcj** [comment]
Any progress on this?  AFAIK there is no way to work with `descriptor.proto` in ruby.

~~~protobuf
syntax = "proto3";
package foo;
import "google/protobuf/descriptor.proto";
~~~

:cry:

**Comment by zanker** [comment]
Sorry! I left Square back in October and the team who was going to take it over got reshuffled so they're not going to have time to pick it back up.

The PR is at https://github.com/google/protobuf/pull/3606, and all the proto side code is pretty solid from the testing I did. The blocker was the proto syntax, which wasn't narrowed down until after I had already left.

If you want to pick up that part of the PR, I'd be happy to walk you through on whatever part of the PR isn't clear.

**Comment by pechorin** [comment]
Is where any plans to implement custom options in ruby? Custom options looks great for expressions some meta logic and it is so sad what ruby does not include this abilitites.

**Comment by ebenoist** [comment]
It looks like there _is_ support in upb.c. What is the thinking around what the syntax would look like? If it's straightforward I can put up a PR.

**Comment by pechorin** [comment]
So, custom options currently cannot be compiled via protoc2, but can be compiled via protoc3 but without any options in compiled ruby pb file. So it will be cool to provide meta options like this in ruby output file:

```proto
syntax = "proto2";

import "google/protobuf/descriptor.proto";

extend google.protobuf.EnumValueOptions {
   optional string name = 1001;
}
message Operation {

  enum ClassName {
    OPERATION = 0 [(name) = 'Operation'];
    PUSH_OPERATION = 1  [(name) = 'PushOperation'];
  }
}
```

So, you ask about how it will be implemented in ruby generated classes? Am i understand you right?

**Comment by ebenoist** [comment]
@pechorin Thats correct, just looking for guidance on the Ruby syntax/methods for the generated classes. It seems like `upb.c` already has support for reading out the options, they would just need to be exposed.

**Comment by krak3n** [comment]
Any news on this? I also wanted to add options to an enum and it works perfectly for go but not ruby 😞 

```proto
syntax = "proto3";

package some.package;

import "google/protobuf/descriptor.proto";

extend google.protobuf.EnumOptions {
  optional int64 default_http_status_code = 50000;
}

extend google.protobuf.EnumValueOptions {
  optional int64 http_status_code = 50001;
}

message Error {
  enum Code {
      option (default_http_status_code) = 400;

      CODE_UNSPECIFIED = 0 [(http_status_code) = 500];
      CODE_INVALID = 1;
      CODE_EXPIRED = 2;
      CODE_NOT_FOUND = 3 [(http_status_code) = 404];
  }

  Code code = 1;
}
```

```
[libprotobuf WARNING ../../third_party/protobuf/src/google/protobuf/compiler/ruby/ruby_generator.cc:512] Omitting proto2 dependency 'google/protobuf/descriptor.proto' from proto3 output file 'banked/consent/v1/errors_pb.rb' because we don't support proto2 and no proto2 types from that file are being used.
```

**Comment by i0jupiter** [comment]
So I have been playing with this more and found that the code generated by `protobuf/releases/download/v3.17.3/protoc-3.17.3-linux-x86_64.zip` is different from `protobuf/releases/download/v3.17.3/protoc-3.17.3-osx-x86_64.zip`!

The linux one doesn't have the `require 'google/protobuf/descriptor_pb'` line, so the above error doesn't exist. Tracking this down has been an incredibly frustrating experience—I hope you choose to do something about it.

This also means that, while extensions are still unavailable in Ruby, at least proto definitions with extensions can be used with Ruby in a multi-lang stack.

**Comment by haberman** [comment]
I'm sorry this has been frustrating to track down.

I was not able to reproduce your report of getting different output from `protoc-3.17.3-linux-x86_64.zip` and `protoc-3.17.3-osx-x86_64.zip`. I tried them both just now and did not get the `require 'google/protobuf/descriptor_pb'` line in either.

If you are getting `require 'google/protobuf/descriptor_pb'` I think you must be using a build of protoc that includes https://github.com/protocolbuffers/protobuf/pull/9003, which was released in https://github.com/protocolbuffers/protobuf/releases/tag/v3.18.1.

I think the root cause of the issue you encountered is a bug in https://github.com/protocolbuffers/protobuf/pull/9003. Specifically, the generated file `descriptor_pb.rb`is throwing an error. This error occurs because `google/protobuf` has not yet been required:

```
google-protobuf-3.18.1-universal-darwin/lib/google/protobuf/descriptor_pb.rb:5:in `<top (required)>': uninitialized constant Google (NameError)
```

We need to change the code generator such that `descriptor_pb.rb` starts with the line `require 'google/protobuf'`. A short-term workaround would be to manually require 'google/protobuf' before importing any other `foo_pb` generated file.

All of this is unrelated to whether Ruby supports custom options or not (just noting this for future readers who come across this bug).

Custom options for Ruby are now within reach; the upb library which powers Ruby supports extensions as of https://github.com/protocolbuffers/upb/pull/421 and custom options as of https://github.com/protocolbuffers/upb/pull/426. Full custom option support for Ruby should be possible now.

**Comment by haberman** [comment]
I opened a separate issue for the problem with `google/protobuf/descriptor_pb`: https://github.com/protocolbuffers/protobuf/issues/9122

This will be fixed in https://github.com/protocolbuffers/protobuf/pull/9121

I misspoke; this was not a bug in https://github.com/protocolbuffers/protobuf/pull/9003, this was a bug in my previous PR https://github.com/protocolbuffers/protobuf/pull/8850

@i0jupiter I'll address your other questions momentarily.

**Comment by seanhuang-stripe** [comment]
@haberman Hey! I've been following along this thread for a while and I might have a personal interest in taking this up, especially given that you mentioned the underlying foundations are ready. What is the best way to receive some guidance from you to get started on this?

**Comment by haberman** [comment]
@seanhuang-stripe thank you for the interest!

There are two main pieces to supporting custom options:

1. Extending the generated code to embed a serialized descriptor, so that all custom options are plumbed to the upb layer. The simplest solution to this would be to simply have all protos use `GenerateBinaryDescriptor()` instead of `GenerateDslDescriptor()`, eg. https://github.com/protocolbuffers/protobuf/blob/125dfd343e823e0b16209520d0c6d2c63f713bb2/src/google/protobuf/compiler/ruby/ruby_generator.cc#L525-L531 However that will also make the generated code less readable, since it would no longer use the DSL, which users may object to. I think we could accommodate that readability concern by generating separate [RBI files](https://sorbet.org/docs/rbi) for the generated code. This could get the best of both worlds: small, fast-loading `_pb.rb` files that preserve all custom options, and larger, readable `_pb.rbi` files that give IDEs, type checkers, and human readers something that represents the fully-expanded interface.  cc @JasonLunn who I've discussed this idea with previously.
2. Extending the runtime to surface the custom options. This would be reasonably straightforward I think, you could adapt the Python implementation which currently supports this eg. https://github.com/protocolbuffers/upb/blob/0e8772fc20e5a0a2fa1f326c79d494374871ef94/python/descriptor.c#L103-L106 There is a question of whether we should return a new instance each time or whether we should return a shared instance that is frozen, we can figure out those details once the basic implementation is in place.

I would be very happy to review a design doc and PRs for this feature.

**Comment by JasonLunn** [comment]
Marking this as blocked on #9495. Until we can make the generated code even uglier without degrading human readability (by adopting RBS with will allow us to separate those concerns), we don't want to do the work in point 1 above.

**Comment by haberman** [comment]
This is unblocked now that https://github.com/protocolbuffers/protobuf/issues/10492 is fixed. More info: https://protobuf.dev/news/2023-04-20/

**Comment by jsteinberg** [comment]
Thank you for the quick response.  Also, that way is much simpler.  I should have some time this week to finish this up and get a PR submitted.

I'm still struggling with reading the custom options.  I'm hoping you can point me in the right direction.

Given the following proto, I am able to get `deprecated` as you showed.
```proto
extend google.protobuf.MessageOptions {
  string custom_message_opt = 7739036;
}

message Event {
  option deprecated = true;
  option (custom_message_opt) = "CUSTOM MESSAGE OPTION";

  string id = 1;
}
```

```ruby
d = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("Event")

d.serialized_options
=> "\x18\x01\xE2\xE9\xC2\x1D\x15CUSTOM MESSAGE OPTION" 

# I'll add to the ruby interface like you suggested.
options = Google::Protobuf::MessageOptions.decode(d.serialized_options)
=> <Google::Protobuf::MessageOptions: deprecated: true, uninterpreted_option: []>

options.deprecated
=> true

options.custom_message_opt
NoMethodError: undefined method `custom_message_opt' for <Google::Protobuf::MessageOptions: deprecated: true, uninterpreted_option: []>:Google::Protobuf::MessageOptions

options.class.encode(options)
=> "\x18\x01\xE2\xE9\xC2\x1D\x15CUSTOM MESSAGE OPTION"
```

Thanks again

**Comment by haberman** [comment]
That is a good question. It looks like we haven't added Ruby APIs to support extensions yet. But the good news is, that should be pretty easy to do also at this point.

We should add the following code to `DescriptorPool_lookup()`:

```c
  fielddef = upb_DefPool_FindExtensionByName(self->symtab, name_str);
  if (fielddef) {
    return get_fielddef_obj(_self, enumdef);
  }
```

Then you could write something like this:

```ruby
ext = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("custom_message_opt")
ext.get(options)
```

We might also want to make `Message.[]` accept a FieldDescriptor, so you could get slightly nicer syntax:

```ruby
options[ext]
```

The final thing we would want to do is add these `lookup` calls to the generated code, so a user could just reference the extension by name, instead of having to call `lookup` first.

But all of that should probably go into a separate PR, which is to add extensions support to Ruby.

By the way, both PRs will need to go through a round of API review internally. But I expect that the APIs I proposed above should be relatively uncontroversial.
